### ISSUE TITLE
Compiler bug, Module validation failed

### ISSUE THREAD

**Issue by Nephos** [crystal-lang/crystal#4858]
Hello,

I got this message today:

```text
make test
...

Module validation failed: !dbg attachment points at wrong subprogram for function
!91 = distinct !DISubprogram(name: "termine", linkageName: "termine", scope: !14, file: !14, line: 24, type: !6, isLocal: true, isDefinition: true, scopeLine: 24, isOptimized: false, unit: !0, variables: !2)
i1 (%"Core::Item::Skill"*, %String*, %"Array(String)"*)* @"*Core::Item::Skill#termine<String, Array(String)>:Bool"
  %0 = call %"Hash(String, Proc(Core::Item::Skill, Array(String), Bool))"** @"~Core::Item::Skill::TERMINE_OPERATORS:read"(), !dbg !93
!93 = !DILocation(line: 73, column: 5, scope: !94)
!94 = distinct !DISubprogram(name: "xp_is_max", linkageName: "xp_is_max", scope: !14, file: !14, line: 72, type: !6, isLocal: true, isDefinition: true, scopeLine: 72, isOptimized: false, unit: !0, variables: !2)
!94 = distinct !DISubprogram(name: "xp_is_max", linkageName: "xp_is_max", scope: !14, file: !14, line: 72, type: !6, isLocal: true, isDefinition: true, scopeLine: 72, isOptimized: false, unit: !0, variables: !2)

0xc9ea8d239e: verify at /home/poulet_a_hdd/Projects/.clone/crystal/src/llvm/module.cr 74:9
0xc9ea8c4da1: finish at /home/poulet_a_hdd/Projects/.clone/crystal/src/compiler/crystal/codegen/codegen.cr 338:9
0xc9e9d12327: codegen at /home/poulet_a_hdd/Projects/.clone/crystal/src/compiler/crystal/codegen/codegen.cr 67:7
0xc9e9d12298: codegen:debug:single_module:expose_crystal_main at /home/poulet_a_hdd/Projects/.clone/crystal/src/compiler/crystal/codegen/codegen.cr 63:5
0xc9ea801de3: codegen at /home/poulet_a_hdd/Projects/.clone/crystal/src/compiler/crystal/compiler.cr 22:7
0xc9ea804553: compile at /home/poulet_a_hdd/Projects/.clone/crystal/src/compiler/crystal/compiler.cr 139:16
0xc9e9cd6d38: spec at /home/poulet_a_hdd/Projects/.clone/crystal/src/compiler/crystal/command/spec.cr 70:14
0xc9e9cd1d8a: run at /home/poulet_a_hdd/Projects/.clone/crystal/src/compiler/crystal/command.cr 93:9
0xc9e9cd168b: run at /home/poulet_a_hdd/Projects/.clone/crystal/src/compiler/crystal/command.cr 49:5
0xc9e9cd164b: run at /home/poulet_a_hdd/Projects/.clone/crystal/src/compiler/crystal/command.cr 48:3
0xc9e9b7d83b: __crystal_main at /home/poulet_a_hdd/Projects/.clone/crystal/src/compiler/crystal.cr 8:1
0xc9e9b92459: main at /home/poulet_a_hdd/Projects/.clone/crystal/src/main.cr 12:15
0x7f6bc9a424ca: __libc_start_main at ??
0xc9e9b7c92a: _start at ??
0x0: ??? at ??

Error: you've found a bug in the Crystal compiler. Please open an issue, including source code that will allow us to reproduce the bug: https://github.com/crystal-lang/crystal/issues
make: *** [Makefile:12: test] Error 1
```

I used `Crystal 0.23.0+25 [8f2a9c2eb] (2017-07-26) LLVM 4.0.1` and `Crystal 0.23.1 (2017-07-17) LLVM 4.0.1`.

I've reduce this code to

```crystal
class Foo
  # Create a list of operators
  CONSTANT_LIST = {
    "bar" => -> (on : Foo) { on.bar() },
  }

  # Read into the list of operators functions to call the right one.
  def dynamic_call(operator : String)
    operator_fct = CONSTANT_LIST[operator]
    operator_fct.call(self)
  end

  def bar
    return "{}"
  end
end

```

with the spec:

```crystal
require "spec"

describe Foo do
  it "basic test" do
    skill = Foo.new().dynamic_call("bar")
  end
end
```

Other information:

my specs are:

    spec/
    ├── core
    │   └── some_spec.cr (in this file)
    ├── Hackware_spec.cr
    └── spec_helper.cr

If I move the specs into Hackware_spec (Hackware is the name of the project) then the bug disappears too.  Same thing if I do remove the "return" in

```crystal
  def bar()
    "{}"
  end
```

**Comment by Nephos** [comment]
I fail to reproduce the bug right now, after I did a lot of tests/modifications and roll back to the bugged version.

Could it be because of the cache of the compiler ?

**Comment by booss** [comment]
Hi, I found a similar issue (Crystal 0.23.1 (2017-09-10) LLVM 5.0.0 on Linux 64bits)

**Code**:

```crystal
class Dummy
    def initialize(@iv1 = "", @iv2 = "")
    end
end

def test(array)
    array.each do |a|
	puts "1: #{a.@iv1} #{a.@iv2}"
	spawn do
	    begin
		raise "dummy"
	    rescue
		puts "2: #{a.@iv1} #{a.@iv2}"
	    end
	end
    end
    Fiber.yield
end

test [Dummy.new("1", "2"),Dummy.new("3", "4"),Dummy.new("5", "6")]
sleep 1
```

**Run**:
crystal file.cr

**Output**:

Module validation failed: Stored value type does not match pointer operand type!
  store i32 1072, i1* %self, !dbg !256
 i1
Error opening file 'crystal' with mode 'r': No such file or directory (Errno)
Failed to raise an exception: END_OF_STACK
[0x5650fe286d17] __crystal_raise +375
[0x5650fe28b625] ???
[0x5650fe29bf0d] ???
[0x5650fe294591] ???
[0x5650fe29395a] ???
[0x5650fe290877] ???
[0x5650fe2ad799] main +13817
[0x7f21204664ca] __libc_start_main +234
[0x5650fe2867ba] _start +42
[0x0] ???

**Comment by akzhan** [comment]
Repetable on master:

```
➜  ~ Projects/separated/crystal/bin/crystal 1.cr
Using compiled compiler at `.build/crystal'
Assertion failed: (getOperand(0)->getType() == cast<PointerType>(getOperand(1)->getType())->getElementType() && "Ptr must be a pointer to Val type!"), function AssertOK, file /tmp/llvm-5.0-20170910-69430-1w13e9i/llvm-5.0.0.src/lib/IR/Instructions.cpp, line 1400.
[1]    15496 abort      Projects/separated/crystal/bin/crystal 1.cr
```

**Comment by Blacksmoke16** [comment]
https://play.crystal-lang.org/#/r/6osh

```cr
abstract struct Command
  macro inherited
    macro finished
      class_getter command : Proc(Nil) = -> do
        ->{ \{{@type}}.execute }.call
      end
    end
  end
end

struct ZZZ < Command
  def self.execute : Nil
    "foo".to_s
  end
end

{{Command.subclasses}}.first.command.call
```

```
Module validation failed: !dbg attachment points at wrong subprogram for function
!4 = distinct !DISubprogram(name: "command", linkageName: "command", scope: !5, file: !5, line: 11, type: !6, isLocal: true, isDefinition: true, scopeLine: 11, isOptimized: false, unit: !0, variables: !2)
%"->" ()* @"*ZZZ::command:Proc(Nil)"
  %0 = call %"->"* @"~ZZZ::command:read"(), !dbg !9
!9 = !DILocation(line: 12, column: 3, scope: !10)
!10 = distinct !DISubprogram(name: "execute", linkageName: "execute", scope: !5, file: !5, line: 12, type: !6, isLocal: true, isDefinition: true, scopeLine: 12, isOptimized: false, unit: !0, variables: !2)
!10 = distinct !DISubprogram(name: "execute", linkageName: "execute", scope: !5, file: !5, line: 12, type: !6, isLocal: true, isDefinition: true, scopeLine: 12, isOptimized: false, unit: !0, variables: !2)
 (Exception)
  from ???
  from ???
  from ???
  from ???
  from ???
  from ???
  from ???
  from __crystal_main
  from main
  from __libc_start_main
  from _start
  from ???
Error: you've found a bug in the Crystal compiler. Please open an issue, including source code that will allow us to reproduce the bug: https://github.com/crystal-lang/crystal/issues
```
Another reproduction.

However changing the return type from `Nil` to `String` fixes the issue?

https://play.crystal-lang.org/#/r/6osi

```cr
abstract struct Command
  macro inherited
    macro finished
      class_getter command : Proc(String) = -> do
        ->{ \{{@type}}.execute }.call
      end
    end
  end
end

struct ZZZ < Command
  def self.execute : String
    "foo"
  end
end

{{Command.subclasses}}.first.command.call
```

**Comment by jwoertink** [comment]
I just ran in to this using crystal 0.34.0
```
[12:44PM] avram (refactors/operation)$ crystal spec
Assertion failed: (getOperand(0)->getType() == cast<PointerType>(getOperand(1)->getType())->getElementType() && "Ptr must be a pointer to Val type!"), function AssertOK, file /var/cache/omnibus/src/llvm/llvm-6.0.1.src/lib/IR/Instructions.cpp, line 1400.
/Users/jeremywoertink/.asdf/installs/crystal/0.34.0/bin/crystal: line 103:  5012 Abort trap: 6           "$INSTALL_DIR/embedded/bin/crystal" "$@"
```

I should note that I use `asdf` to install crystal which just pulls down the binary. I've asked about the llvm 6 deal before, but I forget what the answer was. My preference would be to install from asdf but have crystal built against llvm 10, but I'm not sure how to make that happen (or if that's even the fix).

**Comment by jwoertink** [comment]
oh, this is fun. I just went to LLVM 10, and I get a new error.

```
[13:06PM] avram (refactors/operation)$ crystal spec
Invalid memory access (signal 11) at address 0x0
[0x101b15e97] *CallStack::print_backtrace:Int32 +39
[0x101a5f47a] __crystal_sigfault_handler +522
[0x7fff6c5075fd] _sigtramp +29
[0x102b0d047] _ZN4llvm11PointerType3getEPNS_4TypeEj +21
[0x1026e10cf] _ZN4llvm17GetElementPtrInst16getGEPReturnTypeEPNS_4TypeEPNS_5ValueENS_8ArrayRefIS4_EE +63
[0x1026fe17b] _ZN4llvm17GetElementPtrInst6CreateEPNS_4TypeEPNS_5ValueENS_8ArrayRefIS4_EERKNS_5TwineEPNS_11InstructionE +107
[0x101f39762] _ZN4llvm9IRBuilderINS_14ConstantFolderENS_24IRBuilderDefaultInserterEE17CreateInBoundsGEPEPNS_4TypeEPNS_5ValueENS_8ArrayRefIS7_EERKNS_5TwineE +130
[0x102a80fec] LLVMBuildInBoundsGEP +84
[0x101678ec8] *Crystal::CodeGenVisitor@Crystal::LLVMBuilderHelper#gep<LLVM::Value, Int32, Int32, String>:LLVM::Value +168
[0x1016d9e24] *Crystal::CodeGenVisitor#read_instance_var<Crystal::Type+, Crystal::Type+, String, LLVM::Value>:Bool +100
[0x1016a10b6] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +23542
[0x1016e59ca] *Crystal::CodeGenVisitor#visit<Crystal::Assign>:(Bool | Nil) +298
[0x10169ba71] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1457
[0x1016ea9db] *Crystal::CodeGenVisitor#codegen_cond<Crystal::ASTNode+>:LLVM::Value +27
[0x10169ee56] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +14742
[0x1016b7ff7] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+, Bool, Crystal::CodeGenVisitor::ModuleInfo, Bool, Bool>:LLVM::Function +4711
[0x1016d43a8] *Crystal::CodeGenVisitor#target_def_fun<Crystal::Def+, Crystal::Type+>:LLVM::Function +2664
[0x1016c6718] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +2632
[0x10169ec38] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +14200
[0x1016b7ff7] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+, Bool, Crystal::CodeGenVisitor::ModuleInfo, Bool, Bool>:LLVM::Function +4711
[0x1016d43a8] *Crystal::CodeGenVisitor#target_def_fun<Crystal::Def+, Crystal::Type+>:LLVM::Function +2664
[0x1016c6718] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +2632
[0x10169ec38] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +14200
[0x1016d1b5b] *Crystal::CodeGenVisitor#prepare_call_args_non_external<Crystal::Call, Crystal::Def+, Crystal::Type+>:Tuple(Array(LLVM::Value), Bool) +251
[0x1016c66f0] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +2592
[0x10169ec38] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +14200
[0x10169efc4] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +15108
[0x10169bb49] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1673
[0x1016b7ff7] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+, Bool, Crystal::CodeGenVisitor::ModuleInfo, Bool, Bool>:LLVM::Function +4711
[0x1016d43a8] *Crystal::CodeGenVisitor#target_def_fun<Crystal::Def+, Crystal::Type+>:LLVM::Function +2664
[0x1016c6718] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +2632
[0x10169ec38] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +14200
[0x10169efc4] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +15108
[0x10169bb49] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1673
[0x1016a3618] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +33112 (2 times)
[0x10169f037] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +15223
[0x10169bb49] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1673
[0x1016a3618] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +33112
[0x10169cce6] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +6182
[0x10169d262] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +7586
[0x10169bb49] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1673
[0x1016c6edd] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +4621
[0x10169ec38] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +14200
[0x10169bb49] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1673
[0x1016c6edd] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +4621
[0x10169ec38] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +14200
[0x1016c6edd] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +4621
[0x10169ec38] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +14200
[0x1016b7ff7] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+, Bool, Crystal::CodeGenVisitor::ModuleInfo, Bool, Bool>:LLVM::Function +4711
[0x1016d43a8] *Crystal::CodeGenVisitor#target_def_fun<Crystal::Def+, Crystal::Type+>:LLVM::Function +2664
[0x1016c6718] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +2632
[0x10169ec38] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +14200
[0x10169bb49] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1673
[0x1016b7ff7] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+, Bool, Crystal::CodeGenVisitor::ModuleInfo, Bool, Bool>:LLVM::Function +4711
[0x1016d43a8] *Crystal::CodeGenVisitor#target_def_fun<Crystal::Def+, Crystal::Type+>:LLVM::Function +2664
[0x1016c6718] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +2632
[0x10169ec38] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +14200
[0x10169bb49] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1673
[0x1016b7ff7] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+, Bool, Crystal::CodeGenVisitor::ModuleInfo, Bool, Bool>:LLVM::Function +4711
[0x1016d43a8] *Crystal::CodeGenVisitor#target_def_fun<Crystal::Def+, Crystal::Type+>:LLVM::Function +2664
[0x1016c6718] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +2632
[0x10169ec38] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +14200
[0x10169bb49] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1673
[0x1016b7ff7] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+, Bool, Crystal::CodeGenVisitor::ModuleInfo, Bool, Bool>:LLVM::Function +4711
[0x1016d43a8] *Crystal::CodeGenVisitor#target_def_fun<Crystal::Def+, Crystal::Type+>:LLVM::Function +2664
[0x1016c6718] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +2632
[0x10169ec38] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +14200
[0x1016c2832] *Crystal::CodeGenVisitor#run_instance_vars_initializers<Crystal::Type+, (Crystal::ClassType+ | Crystal::GenericClassType+), LLVM::Value>:Nil +1506
[0x1016c1728] *Crystal::CodeGenVisitor#allocate_aggregate<Crystal::Type+>:LLVM::Value +3656
[0x1016c0410] *Crystal::CodeGenVisitor#codegen_primitive_allocate<Crystal::Primitive+, Crystal::Def+, Array(LLVM::Value)>:LLVM::Value +384
[0x1016da741] *Crystal::CodeGenVisitor#codegen_primitive<Crystal::Call, Crystal::Primitive+, Crystal::Def+, Array(LLVM::Value)>:Nil +2001
[0x1016c62c5] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +1525
[0x10169ec38] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +14200
[0x1016e59ca] *Crystal::CodeGenVisitor#visit<Crystal::Assign>:(Bool | Nil) +298
[0x10169ba71] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1457
[0x10169bb49] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1673
[0x1016b7ff7] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+, Bool, Crystal::CodeGenVisitor::ModuleInfo, Bool, Bool>:LLVM::Function +4711
[0x1016d43a8] *Crystal::CodeGenVisitor#target_def_fun<Crystal::Def+, Crystal::Type+>:LLVM::Function +2664
[0x1016c6718] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +2632
[0x10169ec38] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +14200
[0x1016d1b5b] *Crystal::CodeGenVisitor#prepare_call_args_non_external<Crystal::Call, Crystal::Def+, Crystal::Type+>:Tuple(Array(LLVM::Value), Bool) +251
[0x1016c66f0] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +2592
[0x10169ec38] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +14200
[0x1016b7ff7] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+, Bool, Crystal::CodeGenVisitor::ModuleInfo, Bool, Bool>:LLVM::Function +4711
[0x1016d43a8] *Crystal::CodeGenVisitor#target_def_fun<Crystal::Def+, Crystal::Type+>:LLVM::Function +2664
[0x1016c6718] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +2632
[0x10169ec38] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +14200
[0x1016e59ca] *Crystal::CodeGenVisitor#visit<Crystal::Assign>:(Bool | Nil) +298
[0x10169ba71] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1457
[0x10169bb49] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1673
[0x1016a3618] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +33112
[0x10169bb49] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1673
[0x1016a245f] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +28575
[0x1016c6edd] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +4621
[0x10169ec38] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +14200
[0x1016b7ff7] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+, Bool, Crystal::CodeGenVisitor::ModuleInfo, Bool, Bool>:LLVM::Function +4711
[0x10169d651] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +8593
[0x1016c5fbe] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +750
[0x10169ec38] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +14200
[0x10169bb49] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1673
[0x1016b7ff7] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+, Bool, Crystal::CodeGenVisitor::ModuleInfo, Bool, Bool>:LLVM::Function +4711
[0x10169d651] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +8593
[0x1016c5fbe] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +750
[0x10169ec38] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +14200
[0x10169cce6] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +6182
[0x10169bf54] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +2708
[0x10169bb49] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1673
[0x10164ea39] *Crystal::Compiler#codegen<Crystal::Program, Crystal::ASTNode+, Array(Crystal::Compiler::Source), String>:(Tuple(Array(Crystal::Compiler::CompilationUnit), Array(String)) | Nil) +185
[0x101a798d7] *Crystal::Compiler#compile<Array(Crystal::Compiler::Source), String>:Crystal::Compiler::Result +135
[0x101aecb13] *Crystal::Command#spec:NoReturn +3043
[0x101a6c2ec] *Crystal::Command#run:(Bool | Crystal::Compiler::Result | Nil) +27948
[0x1010c7e64] __crystal_main +6836
[0x1010cd0e1] main +65
```

Here's where it's actually barfing https://github.com/luckyframework/avram/pull/369/files#diff-b5acc3cb5c67dd011373edc33078710aR20

```crystal
  def self.run(*args, **named_args)
    params = Avram::Params.new
    run(params, *args, **named_args) do |operation, value|
      yield operation, value
    end
  end

  def self.run(params : Avram::Paramable, *args, **named_args)
    operation = self.new(params, *args, **named_args)
    operation.before_run
    value = operation.run
    operation.after_run(value)
    yield operation, value
  end
```

If I change that first method to this:

```crystal
  def self.run(*args, **named_args, &block)
    params = Avram::Params.new
    run(params, *args, **named_args, &block)
  end
```

It still doesn't compile, but I get a compiler error instead:

```
 16 | OperationWithNeeds.run(tags: ["one", "two"], id: 3) do |operation, value|
                         ^--
Error: wrong number of block arguments (given 2, expected 0)
```

Crystal 0.34.0 (2020-04-07)

LLVM: 10.0.0
Default target: x86_64-apple-macosx

Also worth noting that I tried crystal from master branch `Crystal 0.35.0-dev [f964ed56d] (2020-04-28)` and still got the same error.

@waj @bcardiff

**Comment by caspiano** [comment]
OP's example now compiles fine https://carc.in/#/r/bbl6
As does @Blacksmoke16's https://carc.in/#/r/bbl8

[booss's example](https://github.com/crystal-lang/crystal/issues/4858#issuecomment-343439929) reduced is still an issue. Looks very similar to what @straight-shoota describes in #10484 

```
class Dummy
  def initialize(@iv1 = ""); end
end

def test(a)
  spawn do
    puts a.@iv1
  end
end

test Dummy.new("1")
```

https://carc.in/#/r/bbl4


I think this can be closed in favour of #10484

**Comment by I3oris** [comment]
1.5.0: I don't reproduce any of these:

```crystal
class Foo
  # Create a list of operators
  CONSTANT_LIST = {
    "bar" => -> (on : Foo) { on.bar() },
  }

  # Read into the list of operators functions to call the right one.
  def dynamic_call(operator : String)
    operator_fct = CONSTANT_LIST[operator]
    operator_fct.call(self)
  end

  def bar
    return "{}"
  end
end

require "spec"

describe Foo do
  it "basic test" do
    skill = Foo.new().dynamic_call("bar")
  end
end
```
https://play.crystal-lang.org/#/r/dprb

```crystal
class Dummy
    def initialize(@iv1 = "", @iv2 = "")
    end
end
 
def test(array)
    array.each do |a|
	puts "1: #{a.@iv1} #{a.@iv2}"
	spawn do
	    begin
		raise "dummy"
	    rescue
		puts "2: #{a.@iv1} #{a.@iv2}"
	    end
	end
    end
    Fiber.yield
end
 
test [Dummy.new("1", "2"),Dummy.new("3", "4"),Dummy.new("5", "6")]
sleep 1
```
https://play.crystal-lang.org/#/r/dprf

```crystal
abstract struct Command
  macro inherited
    macro finished
      class_getter command : Proc(Nil) = -> do
        ->{ \{{@type}}.execute }.call
      end
    end
  end
end
 
struct ZZZ < Command
  def self.execute : Nil
    "foo".to_s
  end
end
 
{{Command.subclasses}}.first.command.call
```
https://play.crystal-lang.org/#/r/dprc

```crystal
class Dummy
  def initialize(@iv1 = ""); end
end

def test(a)
  spawn do
    puts a.@iv1
  end
end

test Dummy.new("1")
```
https://play.crystal-lang.org/#/r/dpr9
